### PR TITLE
Custom marshal attribute

### PR DESCRIPTION
**Pull request by jasongin**
This is a proposal for a custom marshalling attribute and interface. Custom marshalling is not actually implemented yet in the code-generator. The design is mostly modeled after [System.Runtime.InteropServices.MarshalAsAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshalasattribute).

### CODE REVIEW

File: Runtime/JSMarshalAsAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
+
+using System;
+
+namespace NodeApi;
+
+/// <summary>
+/// Specifies how .NET types are marshalled to and from JavaScript values.
+/// </summary>
+[AttributeUsage(
```
Comment (vmoroz) [inline Runtime/JSMarshalAsAttribute.cs:9]: >AttributeUsage [](http://example.com/codeflow?start=1&length=14)

Should we also add fields?

File: Runtime/JSMarshalAsAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    /// </summary>
+    public JSMarshal Value { get; }
+
+    /// <summary>
+    /// When <see cref="Value" /> is <see cref="JSMarshal.CustomMarshaler" />, gets the type
+    /// that handles marshalling. The type must implement <see cref="IJSMarshaler{T}" />,
+    /// where the type argument matches the type being of the property, parameter, or
+    /// return value the <see cref="JSMarshalAsAttribute" /> is applied to.
+    /// </summary>
+    public Type? CustomMarshalType { get; init; }
```
Comment (vmoroz) [inline Runtime/JSMarshalAsAttribute.cs:36]: >Type? [](http://example.com/codeflow?start=11&length=5)

I wonder if we can use the new generic attributes instead. In that case the restriction that the type must inherit from IJSMarshaler can be written as an explicit restriction checked by the C# compiler.

**Review by vmoroz — APPROVED** [review]
:shipit:

File: Runtime/JSMarshalAsAttribute.cs (hunk 2 above)
Comment (jasongin) [reply Runtime/JSMarshalAsAttribute.cs:36]: I thought about that, but we'd still need a non-generic `JSMarshalAsAttribute` to support uses that don't specify a custom marshaller type. So I didn't think it was worth creating another attribute.

File: Runtime/JSMarshalAsAttribute.cs (hunk 1 above)
Comment (jasongin) [reply Runtime/JSMarshalAsAttribute.cs:9]: Later, if and when we support marshalling fields.

Fields are not supported now, but there's no reason they can't be marshalled in almost the same way as properties, along with an optimization for `const` fields. I don't think it's a priority now though.

**Comment by jasongin** [comment]
I'm abandoning this PR because it's not a top priority at the moment, and was mainly meant as a starting point for discussion. I do still plan to implement this idea eventually. Since opening this PR I've done a lot more work related to marshalling, so I think there can be some improvements to this design when I bring it back.

**Pull request closed without merging**